R/calcElecDemandICT.R

```diff
@@ -0,0 +1,207 @@
+      ieaHistorical <- filter(df, .data$scenario == "IEA (Base)",
+                              .data$period <= endOfHistory) %>%
+        mutate(scenario = "history")
+
+      ieaScenario <- filter(df, .data$scenario == "IEA (Base)",
+                            .data$period > endOfHistory)
+
+      # append IEA future data to each SSP scenario
+      ieaScenarioExpanded <- map_dfr(unique(df$scenario[grepl("^SSP", df$scenario)]), function(ssp) {
+        ieaScenario %>%
+          mutate(scenario = ssp)
+      })
+
+      # combine: historical + SSP scenarios + expanded IEA future
+      rbind(
+        ieaHistorical,
+        filter(df, grepl("^SSP", .data$scenario)),
+        ieaScenarioExpanded
+      )
+    })() %>%
+
+    # add country-level weights (number of DCs)
+    rename("regionTarget" = "region") %>%
+    left_join(regionmap %>%
+                select("region" = "CountryCode", "regionTarget" = "Region"),
+              by = "regionTarget",
+              relationship = "many-to-many") %>%
+    left_join(dataDC %>%
+                select("region", "weight" = "value"),
+              by = "region") %>%
+
+    # disaggregate to ISO country level
+    group_by(across(all_of(c("regionTarget", "period", "variable", "scenario")))) %>%
+    mutate(value = .data$value * .data$weight / sum(.data$weight, na.rm = TRUE)) %>%
+    ungroup() %>%
+    select(-"regionTarget", -"weight") %>%
+
+    # calculate ICT (DC + Network) demand using global network share
+    pivot_wider(names_from = "variable", values_from = "value") %>%
+    cross_join(networkShare) %>%
+    mutate(
+      "lower"   = .data[["lower bound dc"]]      * (1 + .data[["lowerShare"]]),
+      "upper"   = .data[["upper bound dc"]]      * (1 + .data[["upperShare"]]),
+      "mean"    = .data[["mean dc"]]             * (1 + .data[["meanShare"]]),
+      "central" = .data[["central estimate dc"]] * (1 + .data[["centralShare"]])
+    ) %>%
+    pivot_longer(cols = -all_of(c("region", "period", "scenario", "unit")),
+                 names_to = "variable",
+                 values_to = "value") %>%
+
+    # convert TWh/yr to EJ/yr
+    filter(.data$variable %in% c("lower", "upper", "mean", "central")) %>%
+    select(-"unit") %>%
+    mutate(value = .data$value * twh2ej,
+           carrier = "elec",
+           enduse = "ict") %>%
+    replace_na(list("value" = 0))
+
+
+  # extrapolate scenario data to 2100 using linear fit over last two periods (here: 2045-50)
+  data <- data %>%
+    group_by(across(all_of(c("region", "scenario", "variable")))) %>%
+    group_modify(~ {
+      periods <- sort(unique(.x$period))
+
+      if (.y$scenario != "history") {
+        # linear extrapolation for future scenarios
+        fitData <- .x %>%
+          filter(.data$period %in% tail(periods, 2))
+
+        model <- lm(value ~ period, data = fitData)
+
+        .x %>%
+          interpolate_missing_periods(seq.int(endOfHistory + 5, 2100, 5)) %>%
```

Comment:
Would be a bit more efficient to move the interpolation before the grouping as it groups itself.